trunk/src/app/srs_app_source.cpp

```diff
@@ -1769,55 +1770,54 @@ srs_error_t SrsLiveSourceManager::fetch_or_create(SrsRequest* r, ISrsLiveSourceH
     SrsLocker(lock);
 
     string stream_url = r->get_stream_url();
-    std::map<std::string, SrsLiveSource*>::iterator it = pool.find(stream_url);
+    std::map< std::string, SrsSharedPtr<SrsLiveSource> >::iterator it = pool.find(stream_url);
 
     if (it != pool.end()) {
-        SrsLiveSource* source = it->second;
+        SrsSharedPtr<SrsLiveSource>& source = it->second;
 
         // we always update the request of resource,
         // for origin auth is on, the token in request maybe invalid,
         // and we only need to update the token of request, it's simple.
         source->update_auth(r);
-        *pps = source;
+        pps = source;
         return err;
     }
-    
-    SrsLiveSource* source = new SrsLiveSource();
+
+    SrsSharedPtr<SrsLiveSource> source = new SrsLiveSource();
     srs_trace("new live source, stream_url=%s", stream_url.c_str());
 
-    if ((err = source->initialize(r, h)) != srs_success) {
-        srs_freep(source);
+    if ((err = source->initialize(source, r, h)) != srs_success) {
         return srs_error_wrap(err, "init source %s", r->get_stream_url().c_str());
     }
     
     pool[stream_url] = source;
-    *pps = source;
+    pps = source;
     return err;
 }
 
-SrsLiveSource* SrsLiveSourceManager::fetch(SrsRequest* r)
+SrsSharedPtr<SrsLiveSource> SrsLiveSourceManager::fetch(SrsRequest* r)
 {
     // Use lock to protect coroutine switch.
     // @bug https://github.com/ossrs/srs/issues/1230
     // TODO: FIXME: Use smaller scope lock.
     SrsLocker(lock);
     
     string stream_url = r->get_stream_url();
-    std::map<std::string, SrsLiveSource*>::iterator it = pool.find(stream_url);
+    std::map< std::string, SrsSharedPtr<SrsLiveSource> >::iterator it = pool.find(stream_url);
 
     if (it == pool.end()) {
-        return NULL;
+        return SrsSharedPtr<SrsLiveSource>(NULL);
```

Comment:
How about design the fetch api in this way?
 `SrsSharedPtr<SrsLiveSource>& SrsLiveSourceManager::fetch(SrsRequest* r)`, return a `SrsSharedPtr` reference, and Introduce a globally `SrsSharedPtr<SrsLiveSource> _SHARED_NULL_LIVE_SOURCE(NULL)`, return it if the null shared ptr is needed.
 
And for `fetch_or_create`, return the `SrsSharedPtr` reference instead of though the reference parameter: `SrsSharedPtr<SrsLiveSource>& SrsLiveSourceManager::fetch_or_create(SrsRequest* r, ISrsLiveSourceHandler* h, srs_error_t& err)`

Reply:
For `fetch_or_create` it may return error, so we should use `**pps` or `& pps` as output parameter.

For `fetch`, a global NULL instance or temporary instance is ok, because it's not a hot path.

Won't fix.